File: python/day10.py
def solve():
    with open("problems/problem10") as f:
        lines = f.readlines()

    prob1 = 0
    autocomplete_scores = []

    braces = {
        '(':')',
        '[':']',
        '{':'}',
        '<':'>'
    }

    for line in lines:
        opened = []

        first_illegal_character = None

        for c in line.strip():
            if c in braces:
                opened.append(c)
            elif len(opened) > 0:
                expected = braces[opened.pop()]
                if c != expected:
                    first_illegal_character = c
                    break

        if first_illegal_character:
            # Corrupted line
            if first_illegal_character == ')':
                prob1 += 3
            elif first_illegal_character == ']':
                prob1 += 57
            elif first_illegal_character == '}':
                prob1 += 1197
            elif first_illegal_character == '>':
                prob1 += 25137
        else:
            # Incomplete line
            score = 0

            for c in reversed(opened):
                score *= 5
                if c == '(':
                    score += 1
                elif c == '[':
                    score += 2
                elif c == '{':
                    score += 3
                elif c == '<':
                    score += 4
            
            autocomplete_scores.append(score)

    print("Problem 1:", prob1)

    autocomplete_scores.sort()
    print("Problem 2:", autocomplete_scores[len(autocomplete_scores) // 2])

if __name__ == '__main__':
    solve()

File: src/day10.rs
const INPUT: &str = include_str!("../problems/problem10");

pub fn solve() -> crate::Result<()> {
    let lines = INPUT.lines();

    let mut prob1 = 0;
    let mut autocomplete_scores = Vec::new();

    let mut opened: Vec<char> = Vec::new();

    for line in lines {
        opened.clear();

        let mut first_illegal_character = None;

        for c in line.chars() {
            if matches!(c, '(' | '[' | '{' | '<') {
                opened.push(c);
            } else if let Some(x) = opened.pop() {
                let expected = match x {
                    '(' => ')',
                    '[' => ']',
                    '{' => '}',
                    '<' => '>',
                    _ => unreachable!()
                };

                if c != expected {
                    first_illegal_character = Some(c);
                    break;
                }
            }
        }

        if let Some(x) = first_illegal_character {
            // Corrupted line
            prob1 += match x {
                ')' => 3,
                ']' => 57,
                '}' => 1197,
                '>' => 25137,
                _ => 0,
            };
        } else {
            // Incomplete line
            let mut score: u64 = 0;

            for &c in opened.iter().rev() {
                score *= 5;
                score += match c {
                    '(' => 1,
                    '[' => 2,
                    '{' => 3,
                    '<' => 4,
                    _ => unreachable!()
                };
            }

            autocomplete_scores.push(score);
        }
    }

    println!("Problem 1: {}", prob1);

    autocomplete_scores.sort();
    println!("Problem 2: {}", autocomplete_scores[autocomplete_scores.len() / 2]);

    Ok(())
}